### PR TITLE
[SPARK-24721][SQL] Extract Python UDFs at the end of optimizer

### PR DESCRIPTION
**Pull request by icexelloss**
## What changes were proposed in this pull request?
In #12127 , we moved the ExtractPythonUDFs rule to the physical phase, while there was another option: do ExtractPythonUDFs at the end of optimizer.

Currently we hit 2 issues when exacting python UDFs at physical phase:

1. It happens after data source v2 strategy, so data source v2 strategy needs to deal with python udfs carefully and adds project to produce unsafe row for python udf. See #22206
2. It happens after file source strategy, so we may keep Python UDF as data filter in FileSourceScanExec and fail the planner when try to extract it later. 

This PR proposes to move ExtractPythonUDFs to the end of optimizer.
## How was this patch tested?
Add a new regression test

### CODE REVIEW

**Comment by icexelloss** [comment]
cc @cloud-fan . Followed your suggestion here: https://issues.apache.org/jira/browse/SPARK-24721?focusedCommentId=16560537&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16560537

**Comment by icexelloss** [comment]
retest please

File: python/pyspark/sql/tests.py [hunk 1]
```diff
@@ -3367,6 +3367,24 @@ def test_ignore_column_of_all_nulls(self):
         finally:
             shutil.rmtree(path)
 
+    def test_datasource_with_udf_filter_lit_input(self):
```
Comment (gatorsmile) [inline python/pyspark/sql/tests.py]: Add another test case for arrow-based pandas udf?

**Comment by gatorsmile** [comment]
@icexelloss Do we face the same issue for DataSourceStrategy?

File: python/pyspark/sql/tests.py (hunk 1 above)
Comment (icexelloss) [reply python/pyspark/sql/tests.py]: Make sense. Will add.

**Comment by icexelloss** [comment]
@gatorsmile Possibly, let me see if I can create a test case

**Comment by icexelloss** [comment]
@gatorsmile Can you advise how to create a df with data source? All my attempts end up triggering FileSourceStrategy not DataSourceStrategy

**Comment by icexelloss** [comment]
I think another way to fix this is  to move the logic to `ExtractPythonUDF` to ignore `FileScanExec` `DataSourceScanExec` and `DataSourceV2ScanExec` instead of changing all three rules. The downside is that if a XScanExec node with a Python UDF pushed filter throws exception somewhere else, we need to fix that too. Not sure which way is better. But either way, it would be good to create test case with data source and data source V2... Would appreciate some advise on how to create such relation in test

**Comment by cloud-fan** [comment]
@icexelloss we can implement a dummy data source v1/v2 at scala side and scan them in PySpark test.

**Comment by HyukjinKwon** [comment]
> we can implement a dummy data source v1/v2 at scala side

There's an example https://github.com/apache/spark/pull/21007 that implement something in Scala and use it in Python side test.

**Comment by icexelloss** [comment]
Thanks @HyukjinKwon and @cloud-fan ! I will take a look

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala [hunk 2]
```diff
@@ -117,15 +117,16 @@ abstract class EvalPythonExec(udfs: Seq[PythonUDF], output: Seq[Attribute], chil
...
       val schema = StructType(dataTypes.zipWithIndex.map { case (dt, i) =>
         StructField(s"_$i", dt)
       })
 
       // Add rows to queue to join later with the result.
       val projectedRowIter = iter.map { inputRow =>
-        queue.add(inputRow.asInstanceOf[UnsafeRow])
-        projection(inputRow)
+        val unsafeRow = projection(inputRow)
+        queue.add(unsafeRow.asInstanceOf[UnsafeRow])
```
Comment (icexelloss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: This is probably another bug I found in testing this - If the input node to EvalPythonExec doesn't produce UnsafeRow, and cast here will fail. 

I found this in testing when I pass in an test data source scan node, which produces GeneralInternalRow, will throw exception here.

I am happy to submit this as a separate patch if people think it's necessary

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala [hunk 3]
```diff
@@ -133,6 +134,9 @@ object ExtractPythonUDFs extends Rule[SparkPlan] with PredicateHelper {
   }
 
   def apply(plan: SparkPlan): SparkPlan = plan transformUp {
+    // SPARK-24721: Ignore Python UDFs in DataSourceScan and DataSourceV2Scan
+    case plan: DataSourceScanExec => plan
```
Comment (icexelloss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: I get rid of the logic previously in `FileSourceStrategy` to exclude PythonUDF in the filter in favor of this fix - I think this fix is cleaner.

File: python/pyspark/sql/tests.py [hunk 4]
```diff
@@ -3367,6 +3367,35 @@ def test_ignore_column_of_all_nulls(self):
...
+        import pandas as pd
+        import numpy as np
+        from pyspark.sql.functions import udf, pandas_udf, lit, col
+
+        path = tempfile.mkdtemp()
+        shutil.rmtree(path)
+        try:
+            self.spark.range(1).write.mode("overwrite").format('csv').save(path)
+            filesource_df = self.spark.read.csv(path)
```
Comment (icexelloss) [inline python/pyspark/sql/tests.py]: @gatorsmile Added tests for file source, data source and data source v2. I might need to move the pandas_udf tests into another tests because of arrow_requirement :(

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala (hunk 2 above)
Comment (icexelloss) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: Ok.. This seems to break existing tests. Need to look into it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala [hunk 5]
```diff
@@ -117,15 +117,18 @@ abstract class EvalPythonExec(udfs: Seq[PythonUDF], output: Seq[Attribute], chil
           }
         }.toArray
       }.toArray
-      val projection = newMutableProjection(allInputs, child.output)
+
+      // Project input rows to unsafe row so we can put it in the row queue
+      val unsafeProjection = UnsafeProjection.create(child.output, child.output)
```
Comment (icexelloss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: This requires some discussion.

This is probably another bug I found in testing this - If the input node to EvalPythonExec doesn't produce UnsafeRow, and cast here will fail. I don't know if we require data sources to produce unsafe rows, if not, then this is a problem.

I also don't know if this will introduce additional copy if input is already UnsafeRow - it seems like UnsafeProject should be smart to skip the copy but I am not sure if it's actually the case

**Comment by icexelloss** [comment]
Tests pass now. This comment https://github.com/apache/spark/pull/22104/files#r210414941 requires some attention. @cloud-fan Do you think this is the right way to handle GenericInternalRow inputs here?

File: python/pyspark/sql/tests.py [hunk 6]
```diff
@@ -3367,6 +3367,33 @@ def test_ignore_column_of_all_nulls(self):
...
+            datasource_df = self.spark.read \
+                .format("org.apache.spark.sql.sources.SimpleScanSource") \
+                .option('from', 0).option('to', 1).load()
+            datasource_v2_df = self.spark.read \
+                .format("org.apache.spark.sql.sources.v2.SimpleDataSourceV2") \
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: This wouldn't work if test classes are not compiled. I think we should better make another test suite that skips the test if the test classes are not existent.

Comment (icexelloss) [reply python/pyspark/sql/tests.py]: @HyukjinKwon I actually am not sure how does pyspark find these classes and how to check the existence, do you have an example?

Comment (icexelloss) [reply python/pyspark/sql/tests.py]: I can probably check try to check the existence of sql/core/target/scala-2.11/test-classes

Comment (icexelloss) [reply python/pyspark/sql/tests.py]: Hmm... I think this is a bit fragile because things like "scala-2.11"  (scala version can change).

Seems a bit over complicated to do this properly, when do we expect pyspark test to run without compiling scala test classes?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala (hunk 5 above)
Comment (icexelloss) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: Friendly ping @cloud-fan. Do you think forcing a unsafeProject here to deal with non-unsafe rows from data sources are correct? Is there a way to know whether the children nodes output unsafe rows so to avoid unnecessary unsafe projection here?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: Ideally all the operators will produce UnsafeRow. If the data source does not produce UnsafeRow, Spark will make sure there will be a project above it to produce UnsafeRow, so we don't need to worry it here and safely assume the input is always UnsafeRow.

Comment (icexelloss) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: Thanks! I will remove this then.

Comment (icexelloss) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: @cloud-fan Sorry, I don't think I am being very clear...

> If the data source does not produce UnsafeRow, Spark will make sure there will be a project
> above it to produce UnsafeRow

I don't think this is happening for datasource V2 right now:

(Code running in pyspark test)
```
datasource_v2_df = self.spark.read \
                .format("org.apache.spark.sql.sources.v2.SimpleDataSourceV2") \
                .load()
result = datasource_v2_df.withColumn('x', udf(lambda x: x, 'int')(datasource_v2_df['i']))
result.show()
```
The code above fails with:
```
Caused by: java.lang.ClassCastException: org.apache.spark.sql.catalyst.expressions.GenericInternalRow cannot be cast to org.apache.spark.sql.catalyst.expressions.UnsafeRow
	at org.apache.spark.sql.execution.python.EvalPythonExec$$anonfun$doExecute$1$$anonfun$5.apply(EvalPythonExec.scala:127)
	at org.apache.spark.sql.execution.python.EvalPythonExec$$anonfun$doExecute$1$$anonfun$5.apply(EvalPythonExec.scala:126)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$GroupedIterator.takeDestructively(Iterator.scala:1074)
```

I think this is an issue with DataSourceV2 that probably should be addressed in another PR (DataSourceV1 works fine). @cloud-fan WDYT?

Comment (icexelloss) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: Created https://jira.apache.org/jira/browse/SPARK-25213

File: python/pyspark/sql/tests.py (hunk 6 above)
Comment (icexelloss) [reply python/pyspark/sql/tests.py]: Added checks to skip the tests if scala tests are not compiled

**Comment by icexelloss** [comment]
@HyukjinKwon I addressed the comments. Do you mind taking a another look?

File: python/pyspark/sql/tests.py (hunk 4 above)
Comment (icexelloss) [reply python/pyspark/sql/tests.py]: Created separate tests for pandas_udf under ScalarPandasUDFTests

**Comment by HyukjinKwon** [comment]
retest this please

File: python/pyspark/sql/utils.py [hunk 7]
```diff
@@ -152,6 +152,22 @@ def require_minimum_pyarrow_version():
...
+    """ Raise Exception if test classes are not compiled
+    """
+    import os
+    try:
+        spark_home = os.environ['SPARK_HOME']
+    except KeyError:
+        raise RuntimeError('SPARK_HOME is not defined in environment')
+
+    test_class_path = os.path.join(
+        spark_home, 'sql', 'core', 'target', 'scala-2.11', 'test-classes')
```
Comment (HyukjinKwon) [inline python/pyspark/sql/utils.py]: Eh, @icexelloss, can we avoid specific version of `scala-2.11` here?

**Comment by HyukjinKwon** [comment]
@icexelloss, why https://github.com/apache/spark/pull/22104/commits/ccb27bb1ab75e33913f37a4dbe84793e6b9ddeec was reverted in this PR? Looks this is the correct approach.

**Comment by HyukjinKwon** [comment]
I mean, the current code will still break partitioned tables:

```
== Physical Plan ==
*(3) Project [_c0#223, pythonUDF0#231 AS v1#226]
+- BatchEvalPython [<lambda>(0)], [_c0#223, pythonUDF0#231]
   +- *(2) Project [_c0#223]
      +- *(2) Filter (pythonUDF0#230 = 0)
         +- BatchEvalPython [<lambda>(0)], [_c0#223, pythonUDF0#230]
            +- *(1) FileScan csv [_c0#223] Batched: false, Format: CSV, Location: InMemoryFileIndex[file:/tmp/tab3], PartitionFilters: [(<lambda>(0) = 0)], PushedFilters: [], ReadSchema: struct<_c0:string>
```

For instance:

```python
from pyspark.sql.functions import udf, lit, col

spark.range(1).selectExpr("id", "id as value").write.mode("overwrite").format('csv').partitionBy("id").save("/tmp/tab3")
df = spark.read.csv('/tmp/tab3')
df2 = df.withColumn('v1', udf(lambda x: x, 'int')(lit(0)))
df2 = df2.filter(df2['v1'] == 0)

df2.explain()
```

**Comment by cloud-fan** [comment]
can we make `ExtractPythonUDFs` a logical plan instead of physical? then all the problems go away since it happens before the data source strategy.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by HyukjinKwon** [comment]
LGTM

**Comment by cloud-fan** [comment]
Just realized that the PR title and description is not updated. @icexelloss can you update them? thanks!

**Comment by icexelloss** [comment]
@cloud-fan Sure! Updated

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala [hunk 8]
```diff
@@ -31,7 +31,8 @@ class SparkOptimizer(
 
   override def defaultBatches: Seq[Batch] = (preOptimizationBatches ++ super.defaultBatches :+
     Batch("Optimize Metadata Only Query", Once, OptimizeMetadataOnlyQuery(catalog)) :+
-    Batch("Extract Python UDF from Aggregate", Once, ExtractPythonUDFFromAggregate) :+
+    Batch("Extract Python UDFs", Once,
+      Seq(ExtractPythonUDFFromAggregate, ExtractPythonUDFs): _*) :+
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala]: but we already have `ExtractPythonUDFFromAggregate` here...
